// xstate/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var utils = require('./utils.js');
var mapState = require('./mapState.js');
var types = require('./types.js');
var actions = require('./actions.js');
var State = require('./State.js');
var Actor = require('./Actor.js');
var StateNode = require('./StateNode.js');
var Machine = require('./Machine.js');
var interpreter = require('./interpreter.js');
var match = require('./match.js');
var schema = require('./schema.js');

var assign = actions.assign,
    send = actions.send,
    sendParent = actions.sendParent,
    sendUpdate = actions.sendUpdate,
    forwardTo = actions.forwardTo,
    doneInvoke = actions.doneInvoke;

exports.matchesState = utils.matchesState;
exports.mapState = mapState.mapState;
Object.defineProperty(exports, 'ActionTypes', {
  enumerable: true,
  get: function () {
    return types.ActionTypes;
  }
});
Object.defineProperty(exports, 'SpecialTargets', {
  enumerable: true,
  get: function () {
    return types.SpecialTargets;
  }
});
exports.actions = actions;
exports.State = State.State;
exports.toActorRef = Actor.toActorRef;
exports.StateNode = StateNode.StateNode;
exports.Machine = Machine.Machine;
exports.createMachine = Machine.createMachine;
exports.Interpreter = interpreter.Interpreter;
Object.defineProperty(exports, 'InterpreterStatus', {
  enumerable: true,
  get: function () {
    return interpreter.InterpreterStatus;
  }
});
exports.interpret = interpreter.interpret;
exports.spawn = interpreter.spawn;
exports.matchState = match.matchState;
exports.createSchema = schema.createSchema;
exports.t = schema.t;
exports.assign = assign;
exports.doneInvoke = doneInvoke;
exports.forwardTo = forwardTo;
exports.send = send;
exports.sendParent = sendParent;
exports.sendUpdate = sendUpdate;
